encoding/ccf/encode.go

```diff
@@ -0,0 +1,1766 @@
+//	#6.225(composite-type-value)
+func (e *Encoder) encodeResourceInterfaceTypeValue(typ *cadence.ResourceInterfaceType, visited ccfTypeIDByCadenceType) error {
+	rawTagNum := []byte{0xd8, CBORTagResourceInterfaceTypeValue}
+	return e.encodeCompositeTypeValue(
+		typ.ID(),
+		nil,
+		typ.Fields,
+		typ.Initializers,
+		visited,
+		rawTagNum,
+	)
+}
+
+// encodeContractInterfaceTypeValue encodes cadence.ContractInterfaceType as
+// language=CDDL
+// contract-interface-type-value =
+//
+//	; cbor-tag-contract-interface-type-value
+//	#6.226(composite-type-value)
+func (e *Encoder) encodeContractInterfaceTypeValue(typ *cadence.ContractInterfaceType, visited ccfTypeIDByCadenceType) error {
+	rawTagNum := []byte{0xd8, CBORTagContractInterfaceTypeValue}
+	return e.encodeCompositeTypeValue(
+		typ.ID(),
+		nil,
+		typ.Fields,
+		typ.Initializers,
+		visited,
+		rawTagNum,
+	)
+}
+
+// encodeCompositeTypeValue encodes composite and interface type values as
+// language=CDDL
+// composite-type-value = [
+//
+//	id: id,
+//	cadence-type-id: cadence-type-id,
+//	; type is only used by enum type value
+//	type: nil / type-value,
+//	fields: [
+//	    + [
+//	        name: tstr,
+//	        type: type-value
+//	    ]
+//	]
+//	initializers: [
+//	    * [
+//	        * [
+//	            label: tstr,
+//	            identifier: tstr,
+//	            type: type-value
+//	        ]
+//	    ]
+//	]
+//
+// ]
+func (e *Encoder) encodeCompositeTypeValue(
+	cadenceTypeID string,
+	typ cadence.Type,
+	fieldTypes []cadence.Field,
+	initializerTypes [][]cadence.Parameter,
+	visited ccfTypeIDByCadenceType,
+	rawTagNum []byte,
+) error {
+	ccfID, ok := visited[cadenceTypeID]
+	if !ok {
+		return fmt.Errorf("CCF type ID not found for composite type value %s", cadenceTypeID)
+	}
+
+	// Encode given tag number indicating cadence type value.
+	err := e.enc.EncodeRawBytes(rawTagNum)
+	if err != nil {
+		return err
+	}
+
+	// Encode CBOR array head with length 5.
+	err = e.enc.EncodeArrayHead(5)
+	if err != nil {
+		return err
+	}
+
+	// element 0: ccf type id as bstr.
+	// It is used to lookup repeated or recursive types within the same encoded type value.
+	err = e.encodeCCFTypeID(ccfID)
+	if err != nil {
+		return err
+	}
+
+	// element 1: cadence type id as tstr.
+	err = e.encodeCadenceTypeID(cadenceTypeID)
+	if err != nil {
+		return err
+	}
+
+	// element 2: type as nil or type-value.
+	// Type is only used by enum type value.
+	if typ == nil {
+		err = e.enc.EncodeNil()
+	} else {
+		err = e.encodeTypeValue(typ, visited)
+	}
+	if err != nil {
+		return err
+	}
+
+	// element 3: fields as array.
+	err = e.encodeFieldTypeValues(fieldTypes, visited)
+	if err != nil {
+		return err
+	}
+
+	// element 4: initializers as array.
+	return e.encodeInitializerTypeValues(initializerTypes, visited)
+}
+
+// encodeFieldTypeValues encodes composite field types as
+// language=CDDL
+//
+//	fields: [
+//	    + [
+//	        name: tstr,
+//	        type: type-value
+//	    ]
```

Comment:
Same here: Should this be "zero or more"? Maybe already a mistake in the spec

```suggestion
//	    * [
//	        name: tstr,
//	        type: type-value
//	    ]
```

Reply:
Done in commits f89eb0615b95c5bb5deeeed7966fa0d6fa98e9d0 and 42d865792c152b0af33991e4ad5d1b6adf796249.  

Codec now matches today's updates to CCF specs.  Changes to CCF specs include:
- https://github.com/fxamacker/ccf_draft/pull/66
- https://github.com/fxamacker/ccf_draft/pull/67